include/ignition/physics/ContactJointProperties.hh

```diff
@@ -0,0 +1,176 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef IGNITION_PHYSICS_CONTACTJOINTPROPERTIES_HH_
+#define IGNITION_PHYSICS_CONTACTJOINTPROPERTIES_HH_
+
+#include <string>
+#include <vector>
+
+#include <ignition/physics/FeatureList.hh>
+#include <ignition/physics/ForwardStep.hh>
+#include <ignition/physics/Geometry.hh>
+#include <ignition/physics/GetContacts.hh>
+#include <ignition/physics/SpecifyData.hh>
+
+namespace ignition
+{
+namespace physics
+{
+/// \brief SetContactJointPropertiesCallbackFeature is a feature for setting the
+/// properties of a contact joint after it is created but before it affects the
+/// forward step.
+class IGNITION_PHYSICS_VISIBLE SetContactJointPropertiesCallbackFeature
+    : public virtual FeatureWithRequirements<ForwardStep>
+{
+  public: template <typename PolicyT> struct FrictionCoeff
+  {
+    typename PolicyT::Scalar frictionCoeff;
+  };
+
+  public: template <typename PolicyT> struct SecondaryFrictionCoeff
+  {
+    typename PolicyT::Scalar secondaryFrictionCoeff;
+  };
+
+  public: template <typename PolicyT> struct RollingFrictionCoeff
+  {
+    typename PolicyT::Scalar rollingFrictionCoeff;
+  };
+
+  public: template <typename PolicyT> struct SecondaryRollingFrictionCoeff
+  {
+    typename PolicyT::Scalar secondaryRollingFrictionCoeff;
+  };
+
+  public: template <typename PolicyT> struct NormalRollingFrictionCoeff
+  {
+    typename PolicyT::Scalar normalRollingFrictionCoeff;
+  };
+
+  public: template <typename PolicyT> struct SlipCompliance
+  {
+    typename PolicyT::Scalar slipCompliance;
+  };
+
+  public: template <typename PolicyT> struct SecondarySlipCompliance
+  {
+    typename PolicyT::Scalar secondarySlipCompliance;
+  };
+
+  public: template <typename PolicyT> struct RestitutionCoeff
+  {
+    typename PolicyT::Scalar restitutionCoeff;
+  };
+
+  public: template <typename PolicyT> struct FirstFrictionalDirection
+  {
+    typename FromPolicy<PolicyT>::template Use<Vector> firstFrictionalDirection;
+  };
+
+  public: template <typename PolicyT> struct ContactSurfaceMotionVelocity
+  {
+    typename FromPolicy<PolicyT>::template Use<Vector>
+      contactSurfaceMotionVelocity;
+  };
+
+  public: template <typename PolicyT> struct ErrorReductionParameter
+  {
+    typename PolicyT::Scalar errorReductionParameter;
+  };
+
+  public: template <typename PolicyT> struct MaxErrorReductionVelocity
+  {
+    typename PolicyT::Scalar maxErrorReductionVelocity;
+  };
+
+  public: template <typename PolicyT> struct MaxErrorAllowance
+  {
+    typename PolicyT::Scalar maxErrorAllowance;
+  };
+
+  public: template <typename PolicyT> struct ConstraintForceMixing
+  {
+    typename PolicyT::Scalar constraintForceMixing;
+  };
+
+  public: template <typename PolicyT> using ContactSurfaceParams =
+    SpecifyData<ExpectData<
```

Comment:
I noticed a significant performance hit when testing https://github.com/ignitionrobotics/ign-gazebo/pull/869 with the `3k_shapes` world. I narrowed it down to the fact that we're a `CompositeData` type here, which seems to be expensive to copy around. Can we use a `struct` with `std::optional` members?

Reply:
I tried changing the `CompositeData` to `struct` and the resulting performance was exactly the same (dropped from 0.6% to 0.17% on my notebook). I'm more inclined to keep the composite data structure as it allows future extensibility, and I'll make sure ign-gazebo only registers a callback when at least one collision wants contact customization.

Reply:
Ok, after doing a more rigorous test, it shows that `struct` provides __slightly__ better results. Baseline on 3k_shapes without this PR is 0.69%. I've already implemented enabling the callback only when needed, but for these tests, I forced the callback to be created even in the 3k_shapes world. With `CompositeData`, I get 0.23%, while with `struct` I get 0.26-0.27%. Which is still more than 50% performance drop, but I think some drop is expected. Based on these tests, I'll go with the `struct`. It also makes the code better readable.